### PR TITLE
Add sync for logging calls to ensure aggregation across devices

### PR DESCRIPTION
**Pull request by SeanNaren**
# What does this PR do ?

After speaking to @nithinraok a few places do not have sync calls when logging (some are taken care off by the underlying Metrics, or they are already synced across processes like global steps).

I've been a bit cautious and not touched the return objects from `validation_step` or `test_step`. There seems to be some parsing of logic in the ModelPT file so need to investigate further as to what to do with those metrics, and how they are logged.

# Changelog 
- Add `sync_dist` to logging metric calls to ensure aggregation across all devices.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

**Comment by SeanNaren** [comment]
@nithinraok I think the returning dicts that look like `return {'log'{...}}` are fine because of the logic that happens here: https://github.com/NVIDIA/NeMo/blob/main/nemo/core/classes/modelPT.py#L723

Both `asr_model.py` and `classification_model.py` implement `multi_validation_epoch_end`  and `multi_test_epoch_end` to do the aggregations before being logged. I think this is fine, it could be cleaned up slightly but I think the code looks acceptable.

File: nemo/collections/asr/models/label_models.py [hunk 1]
```diff
@@ -307,15 +307,15 @@ def training_step(self, batch, batch_idx):
...
+        self.log('loss', loss, sync_dist=True)
         self.log('learning_rate', self._optimizer.param_groups[0]['lr'])
         self.log('global_step', self.trainer.global_step)
 
         self._accuracy(logits=logits, labels=labels)
         top_k = self._accuracy.compute()
         self._accuracy.reset()
         for i, top_i in enumerate(top_k):
-            self.log(f'training_batch_accuracy_top@{i}', top_i)
+            self.log(f'training_batch_accuracy_top@{i}', top_i, sync_dist=True)
```
Comment (SeanNaren) [inline nemo/collections/asr/models/label_models.py:318]:
```suggestion
            self.log(f'training_batch_accuracy_top@{i}', top_i)
```

Removing this as the metric handles aggregation.

**Comment by titu1994** [comment]
Those older multi stuff needs some cleanup, agreed.

**Comment by titu1994** [comment]
Be careful of adding too many syncs, they can cause deadlocks on the clusters (at least on other domains some versions before)

**Comment by SeanNaren** [comment]
Thanks @titu1994! 

Yeah great point, I need to double check the warnings @nithinraok was mentioning around the calls when `sync_dist` is missing. 

A part of me thinks (atleast for training) it's worth omitting sync for training steps (since sync happens every step) to reduce communication overhead + its usually used to track overall convergence. I do think keeping the sync for validation steps is the right thing to do, but curious what people suggest!

File: nemo/collections/asr/models/classification_models.py [hunk 2]
```diff
@@ -488,7 +488,7 @@ def training_step(self, batch, batch_nb):
         logits = self.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
         loss_value = self.loss(logits=logits, labels=labels)
 
-        self.log('train_loss', loss_value)
+        self.log('train_loss', loss_value, sync_dist=True)
```
Comment (SeanNaren) [inline nemo/collections/asr/models/classification_models.py:491]:
```suggestion
        self.log('train_loss', loss_value)
```

File: nemo/collections/asr/models/classification_models.py [hunk 3]
```diff
@@ -690,14 +690,16 @@ def training_step(self, batch, batch_idx):
...
-            'learning_rate': self._optimizer.param_groups[0]['lr'],
-        }
+        self.log_dict(
+            {
+                'train_loss': loss,
+                'train_mse': train_mse,
+                'train_mae': train_mae,
+                'learning_rate': self._optimizer.param_groups[0]['lr'],
+            },
+            sync_dist=True,
```
Comment (SeanNaren) [inline nemo/collections/asr/models/classification_models.py:700]:
```suggestion
```

**Comment by SeanNaren** [comment]
Dropped the training sync_dist, let's keep it for eval steps as this could vary the loss quite a bit.

File: nemo/collections/asr/models/classification_models.py [hunk 4]
```diff
@@ -690,14 +690,15 @@ def training_step(self, batch, batch_idx):
...
         train_mae = self._mae(preds=logits, target=targets)
 
-        tensorboard_logs = {
-            'train_loss': loss,
-            'train_mse': train_mse,
-            'train_mae': train_mae,
-            'learning_rate': self._optimizer.param_groups[0]['lr'],
-        }
+        self.log_dict(
+            {
```
Comment (titu1994) [inline nemo/collections/asr/models/classification_models.py:694]: You'll lose multi data loader support for classification models with this (GSC has val and test sets)

**Comment by nithinraok** [comment]
@SeanNaren are you keeping or removing `sync_dict` in training step methods?

File: nemo/collections/asr/models/classification_models.py (hunk 4 above)
Comment (SeanNaren) [reply nemo/collections/asr/models/classification_models.py:694]: Thanks @titu1994 I might be missing something here, all I've done is assignment of the variable and combined it into the `self.log_dict` call (primarily because the idea of `tensorboard_logs` has been deprecated for a while in Lightning).

**Comment by SeanNaren** [comment]
> @SeanNaren are you keeping or removing `sync_dict` in training step methods?

hey @nithinraok, I'll remove them based on trying to prevent too much communication overhead per step. 

The PossibleUserWarning shouldn't appear for the training step as the metric is logged every step instead of across the entire epoch.

**Comment by nithinraok** [comment]
Thanks Sean, I see `sync_dict=True` in label_model.py training_step, does it cause communication overhead?

**Review by nithinraok — APPROVED** [review]
